File: Heap/2336. Smallest Number in Infinite Set.cpp
class SmallestInfiniteSet {
public:
    set<int>st;
    int cur = 0;
    SmallestInfiniteSet() {
        
    }
    
    int popSmallest() {
        if(st.empty())
            return ++cur;
        else{
            int res = *st.begin();
            st.erase(res);
            return res;
        }
        
    }
    
    void addBack(int num) {
        if (num <= cur ) {
            st.insert(num);
        } 
    }
};


class SmallestInfiniteSet {
public:
    int cur = 1;
    set<int> added; 
    int popSmallest() {
        if (!added.empty()) {
            int res = *begin(added);
            added.erase(begin(added));
            return res;
        }
        return cur++;
    }
    void addBack(int num) {
        if (num < cur)
            added.insert(num);
    }
};

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * SmallestInfiniteSet* obj = new SmallestInfiniteSet();
 * int param_1 = obj->popSmallest();
 * obj->addBack(num);
 */

File: Heap/2336. Smallest Number in Infinite Set.py
class SmallestInfiniteSet:

    def __init__(self):
        self.cur = 1
        self.heap = []
        self.set = set()

    def popSmallest(self) -> int:
        if self.heap:
            num = heapq.heappop(self.heap)
            self.set.remove(num)
            return num
        self.cur += 1
        return self.cur - 1

    def addBack(self, num: int) -> None:
        if num < self.cur and num not in self.set:
            self.set.add(num)
            heapq.heappush(self.heap, num)


# Your SmallestInfiniteSet object will be instantiated and called as such:
# obj = SmallestInfiniteSet()
# param_1 = obj.popSmallest()
# obj.addBack(num)